### PR TITLE
Wip

### PR DESCRIPTION
**Pull request by jmburges**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  * an [HTTP verb][rfc_http_methods], like `GET`, `PUT`, or `POST`
+  * a path
+  * various headers
+
+HTTP servers are typically stateless. Your Rails server has access to a database, but the request needs to have information about what database rows to look up.
+
+For example, `GET` requests usually encode this information in the path. When you write a route matching `/items/:item_id`, you are telling Rails to pull the value `id` from the request path and save it in `params[:id]`. In your `items_controller`, you'll probably have a method that looks something like:
+
+    def show
+      @item = Item.find!(params[:item_id])
```
Comment (jmburges) [inline README.md]: We generally use just `find` everywhere else in the curriculum

File: README.md [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+Remember what's included in an HTTP request:
+  * an [HTTP verb][rfc_http_methods], like `GET`, `PUT`, or `POST`
+  * a path
+  * various headers
+
+HTTP servers are typically stateless. Your Rails server has access to a database, but the request needs to have information about what database rows to look up.
+
+For example, `GET` requests usually encode this information in the path. When you write a route matching `/items/:item_id`, you are telling Rails to pull the value `id` from the request path and save it in `params[:id]`. In your `items_controller`, you'll probably have a method that looks something like:
+
+    def show
```
Comment (jmburges) [inline README.md]: also wrap these in triple backticks?

File: README.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      @item = Item.find!(params[:item_id])
+    end
+
+Which loads the row for that item from the database and returns it as an ActiveRecord model object, which your `show.html.erb` then renders.
+
+If we want to be able to retrieve the current cart, we need to its id to be somewhere in the HTTP request. Specifically, it must be in the path or the headers.
+
+It would be possible, though quite convoluted, to store this information in the path. This would have strange effects: since the path is shown in the browser's URL bar, a user who copies a URL and sends it to a friend ("check out this neat skirt!") would also be copying their shopping cart ID. Upon loading the page, the friend would
```
Comment (jmburges) [inline README.md]: Great point! I would also mention that the fundamental issue is that some objects only belong to certain users.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+see what's in the user's cart. That is probably not what we want.
+
+Cookies allow us to store this information in the only other place available to us: HTTP headers.
```
Comment (jmburges) [inline README.md]: Available to us, and doesn't won't get accidentally copy pasted to another user

File: README.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+      Set-Cookie: SID=31d4d96e407aad42
+
+      == User Agent -> Server ==
+      Cookie: SID=31d4d96e407aad42
+
+In this example, the server responds to a request with the `Set-Cookie` header. This header sets the value of the `SID` cookie to `31d4d96e407aad42`.
+
+When the user visits another page on the same server, the browser sends the cookie back to the server, including the `Cookie: SID=31d4d96e407aad42` header in its request.
+
+Cookies are stored in the browser. The browser doesn't care about what's in the cookies you set. It just stores the data and sends it along on future requests to your server.
```
Comment (jmburges) [inline README.md]: "You can think of them as a `hash` that just gets stored by the browser. Then when your server asks for the information again, the browser sends it back."

File: README.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+If we want to be able to retrieve the current cart, we need to its id to be somewhere in the HTTP request. Specifically, it must be in the path or the headers.
+
+It would be possible, though quite convoluted, to store this information in the path. This would have strange effects: since the path is shown in the browser's URL bar, a user who copies a URL and sends it to a friend ("check out this neat skirt!") would also be copying their shopping cart ID. Upon loading the page, the friend would
+see what's in the user's cart. That is probably not what we want.
+
+Cookies allow us to store this information in the only other place available to us: HTTP headers.
+
+## What's a cookie, anyway?
+
+Let's see what [the spec][rfc_cookies] has to say:
```
Comment (jmburges) [inline README.md]: We generally shy away from the super technical speak of specs and such. I think it's fine to keep this in here if you'd like, but then after this I'd mention that we don't expect beginners to be able to understand everything. Then go into explaining at how it actually works.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+Their example is quite simple:
+
+== Server -> User Agent ==
+      Set-Cookie: SID=31d4d96e407aad42
+
+      == User Agent -> Server ==
+      Cookie: SID=31d4d96e407aad42
+
+In this example, the server responds to a request with the `Set-Cookie` header. This header sets the value of the `SID` cookie to `31d4d96e407aad42`.
```
Comment (jmburges) [inline README.md]: What would be even cooler is to sort of teach students how to read that technical specs. Maybe have a sentence of the spec and then "translate" it into something a bit more english-y?

File: README.md [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+You can also edit cookies, for example with [this extension|edit_this_cookie].
+
+This presents a problem for us. If users can edit their `cart_id` cookie, then they can see other users' shopping carts.
+
+## Rails to the rescue
+
+Fortunately, Rails has a solution to this. When you set cookies in Rails, you usually don't manipulate the HTTP headers directly. Instead, you use the `session` method. The `session` method is available anywhere in the Rails response cycle, and it behaves like a hash:
+
+    # set cart_id
```
Comment (jmburges) [inline README.md]: triple back ticks

File: README.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    session[:cart_id] = @cart.id
+
+    # load the cart referenced in the session
+    @cart = session[:cart_id]
+
+You can store any simple Ruby object in the session. In fact, we don't need a `Cart` model at all—we can just store a list of items in the session!
+
+Rails manages all session data in a single cookie, named `_YOUR_RAILS_APP_NAME_session`. It *serializes* your all the key/value pairs you set with `session`, turning them into a big string. Whenever you set a `key` with the `session` method, Rails updates the value of its session cookie to this big string.
```
Comment (jmburges) [inline README.md]: Call out that serialize means to convert from a ruby object to a string representation

File: README.md [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+      return signature
+    end
+
+It's guaranteed that given the same message and key, sign will produce output. Also, without the key, it is practically impossible to know what `sign` would return for a given message. That is, signatures can't be forged.
+
+Rails creates a signature for every cookie it sets, and appends the signature to the cookie.
+
+When it receives a cookie, Rails verifies that the signature matches the content (that is, that `sign(cookie_body, secret_key_base) == cookie_signature`).
+
+This prevents cookie tampering. If a user tries to edit their cookie and change the `cart_id`, they signature won't match, and Rails will silently ignore the cookie and set a new one.
```
Comment (jmburges) [inline README.md]: Just mention that, if all of the cryptographic magic doesn't make sense, that's ok. What they need to know is that the signature ensures nobody tampered with the cookie.

Comment (jmburges) [inline README.md]: Don't want people just re-reading this and googling and all that trying to learn how crpytographic signing works haha

File: README.md [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+# Tying it together
+
+In our `items_controller.rb`, we might have an `add_to_cart` method, which is called
```
Comment (jmburges) [inline README.md]: As much as possible, we want students to be able to play with the code that is talked about in the readings. I would include in this repo a rails app that has this code in it. Or it ahs the other bits, and then ask the student to copy paste this code into the controller

Comment (jmburges) [inline README.md]: Yeah I would have them copy paste it into the rails app, and then write a test that proves that they did it

File: README.md [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    # Routed from POST /items/:id/add_to_cart
+    def add_to_cart
+      # Get the item from the path
+      @item = Item.find(params[:id])
+      
+      # Load the cart from the session, or create a new empty cart.
+      current_cart << @item.id
+    end
+
+This way, we can use `current_cart` in our views and layouts too. For example, we may want to show the user how many items are in their cart as part of the layout.
```
Comment (jmburges) [inline README.md]: I would include an examples, showing them how to then use the `session` to display stuff

File: README.md [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+
+## Conclusion
+
+Cookies are foundational for the modern web.
+
+You've probably heard a lot about cookies from non-technical sources. In fact, the EU recently passed a law requiring sites that use cookies to request user consent before doing so.
+
+The effect of this law was not unlike the effect of California's [Proposition 65][prop_65], which states that buildings must have clear signage if they contain any one of a list of chemicals known to the state of California to cause cancer.
```
Comment (jmburges) [inline README.md]: I don't want to bring politics into this as much as possible. Just skip to "most site use cookies..."

File: README.md [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+
+But cookies just let you store data on in a user's browser, so by nature, they can be used for more controversial endeavors. 
+
+For example, Google AdWords sets a cookie and uses that cookie to track what ads you've seen and which ones you've clicked on. The tracking information helps AdWords decide what ads to show you.
+
+This is why, if you click on an ad, you may find that the ad follows you around the Internet. It turns out that this behavior is as effective as it is annoying: people are far more likely to buy things from ads that they've clicked on once.
+
+This use of cookies worries people. And so now the EU has a law.
+
+Cookies, like any technology, are a tool. In the rest of this unit, we're going to be using them to let users log in. Whether you later want to use them in such a way that the EU passes another law is up to you.
```
Comment (jmburges) [inline README.md]: I think reminding students to be aware of what they are saving and to do right by their users is important.

**Comment by jmburges** [comment]
I made a pr for this work. much easier to work in. Back at you!

**Comment by queerviolet** [comment]
I don't understand why that branch push closed the PR :(

**Pull request closed without merging**